File: proyecto_java_estatuto/src/main/java/co/edu/udistrital/estatuto/controlador/FacultadControlador.java
package co.edu.udistrital.estatuto.controlador;

import co.edu.udistrital.estatuto.modelo.Facultad;
import co.edu.udistrital.estatuto.persistencia.JsonDAO;
import co.edu.udistrital.estatuto.util.Validador;
import com.google.gson.reflect.TypeToken;

import java.io.IOException;
import java.lang.reflect.Type;
import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

/**
 * Controlador para la gestión de Facultades.
 *
 * <p>Orquesta las operaciones CRUD sobre {@link Facultad}, aplica validaciones
 * de negocio y delega la persistencia al {@link JsonDAO}. Actúa como
 * intermediario entre la vista y el modelo.</p>
 *
 * @author Roberto Albeiro Pava-Díaz
 * @version 1.0
 */
public class FacultadControlador {

    /** DAO para persistencia de Facultades en JSON. */
    private final JsonDAO<Facultad, String> dao;

    /** Ruta por defecto del archivo de datos. */
    private static final String RUTA_DATOS = "datos/facultades.json";

    /**
     * Construye el controlador con la ruta de datos por defecto.
     */
    public FacultadControlador() {
        Type tipo = new TypeToken<List<Facultad>>() {}.getType();
        this.dao = new JsonDAO<>(RUTA_DATOS, tipo, Facultad::getCodigo);
    }

    /**
     * Construye el controlador con una ruta de datos personalizada.
     * @param rutaDatos ruta al archivo JSON de datos
     */
    public FacultadControlador(String rutaDatos) {
        Type tipo = new TypeToken<List<Facultad>>() {}.getType();
        this.dao = new JsonDAO<>(rutaDatos, tipo, Facultad::getCodigo);
    }

    /**
     * Crea una nueva Facultad con los datos básicos.
     * Valida que el código y el nombre no estén vacíos y que el
     * código no esté ya registrado.
     *
     * @param codigo código institucional (no nulo, no vacío)
     * @param nombre nombre oficial de la facultad
     * @param decano nombre del decano o decana
     * @return la Facultad creada
     * @throws IllegalArgumentException si el código o nombre son inválidos
     * @throws IOException              si ocurre un error de persistencia
     */
    public Facultad crear(String codigo, String nombre, String decano)
            throws IOException {
        if (!Validador.esCodigoValido(codigo)) {
            throw new IllegalArgumentException(
                "Código de facultad inválido: '" + codigo + "'.");
        }
        if (!Validador.noEsVacia(nombre)) {
            throw new IllegalArgumentException("El nombre de la facultad no puede estar vacío.");
        }
        Facultad f = new Facultad(
            Validador.normalizarNombre(nombre),
            Validador.normalizarCodigo(codigo),
            LocalDate.now(),
            Validador.normalizarNombre(decano)
        );
        dao.guardar(f);
        return f;
    }

    /**
     * Crea o actualiza una Facultad (conveniente para la vista).
     * Si la Facultad ya existe, la actualiza; si no, la crea.
     *
     * @param codigo código institucional
     * @param nombre nombre oficial
     * @param decano nombre del decano
     * @return la Facultad creada o actualizada
     */
    public Facultad crearOActualizar(String codigo, String nombre, String decano) {
        try {
            Optional<Facultad> existente = dao.buscarPorId(
                Validador.normalizarCodigo(codigo));
            if (existente.isPresent()) {
                Facultad f = existente.get();
                f.setNombre(Validador.normalizarNombre(nombre));
                f.setDecano(Validador.normalizarNombre(decano));
                dao.actualizar(f);
                return f;
            } else {
                return crear(codigo, nombre, decano);
            }
        } catch (IOException e) {
            System.err.println("Error de persistencia: " + e.getMessage());
            // En caso de error de IO, retorna el objeto sin persistir
            return new Facultad(nombre, codigo, LocalDate.now(), decano);
        }
    }

    /**
     * Retorna todas las facultades registradas.
     *
     * @return lista de facultades (puede estar vacía)
     */
    public List<Facultad> listarTodos() {
        try {
            return dao.buscarTodos();
        } catch (IOException e) {
            System.err.println("Error al listar facultades: " + e.getMessage());
            return new ArrayList<>();
        }
    }

    /**
     * Busca una facultad por su código.
     *
     * @param codigo código de la facultad
     * @return {@link Optional} con la facultad, o vacío si no existe
     */
    public Optional<Facultad> buscarPorCodigo(String codigo) {
        try {
            return dao.buscarPorId(Validador.normalizarCodigo(codigo));
        } catch (IOException e) {
            System.err.println("Error al buscar facultad: " + e.getMessage());
            return Optional.empty();
        }
    }

    /**
     * Actualiza los datos de una facultad existente.
     *
     * @param facultad facultad con datos actualizados
     * @throws IOException si ocurre un error de persistencia
     */
    public void actualizar(Facultad facultad) throws IOException {
        dao.actualizar(facultad);
    }

    /**
     * Elimina una facultad por su código.
     *
     * @param codigo código de la facultad a eliminar
     * @return {@code true} si se eliminó; {@code false} si no existía
     */
    public boolean eliminar(String codigo) {
        try {
            return dao.eliminar(Validador.normalizarCodigo(codigo));
        } catch (IOException e) {
            System.err.println("Error al eliminar facultad: " + e.getMessage());
            return false;
        }
    }

    /**
     * Busca facultades cuyo nombre contiene el término dado (búsqueda parcial).
     *
     * @param termino término de búsqueda (ignorar mayúsculas/minúsculas)
     * @return lista de facultades que coinciden
     */
    public List<Facultad> buscarPorNombre(String termino) {
        try {
            String t = termino.toLowerCase().trim();
            return dao.buscarTodos().stream()
                      .filter(f -> f.getNombre().toLowerCase().contains(t))
                      .toList();
        } catch (IOException e) {
            System.err.println("Error en búsqueda: " + e.getMessage());
            return new ArrayList<>();
        }
    }
}

File: proyecto_rust_estatuto/src/controlador/facultad_controlador.rs
//! # FacultadControlador — capa de control para Facultades
//!
//! Gestiona las operaciones CRUD sobre las `Facultad`es del sistema,
//! coordinando entre la capa de dominio y la de persistencia.
//!
//! ## Equivalencia Java → Rust
//! La clase `FacultadControlador` de Java usa composición con un `JsonDAO<Facultad>`.
//! En Rust, se usa composición con un `JsonDao<Facultad>` que implementa
//! el trait `GenericDao<Facultad, String>`.

use crate::excepcion::errores::EstatutoResult;
use crate::modelo::facultad::Facultad;
use crate::persistencia::dao::GenericDao;
use crate::persistencia::json_dao::JsonDao;

/// Controlador para la gestión de Facultades.
///
/// Provee métodos de negocio que usan el DAO para persistir y recuperar
/// entidades `Facultad`.
pub struct FacultadControlador {
    dao: JsonDao<Facultad>,
}

impl FacultadControlador {
    /// Crea un nuevo controlador usando el archivo JSON en la ruta dada.
    pub fn nuevo(ruta_datos: &str) -> Self {
        let ruta = format!("{}/facultades.json", ruta_datos);
        FacultadControlador {
            dao: JsonDao::nuevo(ruta, |f: &Facultad| f.base.codigo.clone()),
        }
    }

    /// Lista todas las facultades registradas en el sistema.
    pub fn listar_todas(&self) -> EstatutoResult<Vec<Facultad>> {
        self.dao.buscar_todos()
    }

    /// Registra una nueva Facultad en el sistema.
    ///
    /// # Errores
    /// Retorna error si ya existe una facultad con el mismo código.
    pub fn registrar(&mut self, facultad: Facultad) -> EstatutoResult<()> {
        self.dao.guardar(facultad)
    }

    /// Busca una Facultad por su código institucional.
    pub fn buscar(&self, codigo: &str) -> EstatutoResult<Option<Facultad>> {
        self.dao.buscar_por_id(&codigo.to_uppercase())
    }

    /// Actualiza los datos de una Facultad existente.
    pub fn actualizar(&mut self, facultad: Facultad) -> EstatutoResult<()> {
        self.dao.actualizar(facultad)
    }

    /// Elimina una Facultad del sistema por su código.
    ///
    /// # Retorna
    /// `true` si fue encontrada y eliminada, `false` si no existía.
    pub fn eliminar(&mut self, codigo: &str) -> EstatutoResult<bool> {
        self.dao.eliminar(&codigo.to_uppercase())
    }

    /// Busca facultades cuyo nombre contenga el término dado (búsqueda parcial).
    pub fn buscar_por_nombre(&self, termino: &str) -> EstatutoResult<Vec<Facultad>> {
        let todas = self.dao.buscar_todos()?;
        let termino_lower = termino.to_lowercase();
        Ok(todas
            .into_iter()
            .filter(|f| f.base.nombre.to_lowercase().contains(&termino_lower))
            .collect())
    }

    /// Verifica si existe una Facultad con el código dado.
    pub fn existe(&self, codigo: &str) -> EstatutoResult<bool> {
        Ok(self.buscar(codigo)?.is_some())
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::modelo::facultad::Facultad;
    use chrono::NaiveDate;
    use std::env;

    fn controlador_temporal() -> (FacultadControlador, std::path::PathBuf) {
        let dir = env::temp_dir().join(format!("ctrl_fac_{}", uuid::Uuid::new_v4()));
        std::fs::create_dir_all(&dir).unwrap();
        let ctrl = FacultadControlador::nuevo(dir.to_str().unwrap());
        (ctrl, dir)
    }

    fn facultad(codigo: &str) -> Facultad {
        Facultad::nueva(
            &format!("Facultad {}", codigo),
            codigo,
            NaiveDate::from_ymd_opt(2025, 1, 1).unwrap(),
            "Dir",
            "Decano",
        )
        .unwrap()
    }

    #[test]
    fn registrar_y_listar() {
        let (mut ctrl, dir) = controlador_temporal();
        ctrl.registrar(facultad("FAC-01")).unwrap();
        let todas = ctrl.listar_todas().unwrap();
        assert_eq!(todas.len(), 1);
        std::fs::remove_dir_all(dir).ok();
    }

    #[test]
    fn buscar_existente() {
        let (mut ctrl, dir) = controlador_temporal();
        ctrl.registrar(facultad("FAC-02")).unwrap();
        assert!(ctrl.buscar("FAC-02").unwrap().is_some());
        std::fs::remove_dir_all(dir).ok();
    }

    #[test]
    fn eliminar_facultad() {
        let (mut ctrl, dir) = controlador_temporal();
        ctrl.registrar(facultad("FAC-03")).unwrap();
        assert!(ctrl.eliminar("FAC-03").unwrap());
        std::fs::remove_dir_all(dir).ok();
    }
}
